src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -63,14 +64,23 @@ public override void AdvanceTo(SequencePosition consumed)
 
         public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
         {
-            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, examined).Length;
+            var newlyExamined = dataLength - _totalExamined;
             _context.RequestBodyPipe.Reader.AdvanceTo(consumed, examined);
-            OnDataRead(dataLength);
+
+            if (newlyExamined > 0)
+            {
+                OnDataRead(dataLength);
```

Comment:
Nit: Rename to OnDataExamined.

Also, when you fix the newlyExamined logic, you'll want to pass in newlyExamined instead of dataLength.